src/compiler/parser.ts

```diff
@@ -8054,6 +8059,14 @@ namespace Parser {
         return nextToken() === SyntaxKind.SlashToken;
     }
 
+    function isInvalidConditionalType() {
+        if (token() === SyntaxKind.QuestionToken && !(contextFlags & NodeFlags.JavaScriptFile)) {

```

Comment:
This is going to incur a lot of speculative parsing, isn't it? Specifically, every conditional type is going to speculatively `parseType`. Then it'll fail and proceed with normal parsing, which will `parseType` in the exact same location.
It might be possible to salvage either by (1) stopping the speculation before `parseType` or (2) making the speculative parse succeed when the type *is* conditional, and return some parsed portion of the conditional type node. (1) is likely to be lossy, but maybe not as much as dropping the speculation entirely. (2) would require passing the conditional part of the type down into `parseType`. 

If those are too complex, I don't think it's worthwhile -- I reverted this change and it only breaks when the token following `?` could be the start of a type: `{` -- but `,` and `)` are fine. Improving *some* of the errors is good enough in my opinion.

(Reverting this will require splitting the nullable type test into two parts because syntax errors block semantic errors.)

Reply:
> ? could be the start of a type: {

Yes, this is only related to `{`, as it can be the start of a block or the start of a type. I've removed this speculative parsing, and updated test.